scripts/set-react-version.js

```diff
@@ -98,35 +40,192 @@ function keys(obj) {
+      if (code !== 0 || result.length === 0) {
+        resolve({
+          version: undefined,
+          dependencies: {},
+          peerDependencies: {},
+        });
+      } else {
+        const list = JSON.parse(result.join(""));
+        const latest = Array.isArray(list) ? list[list.length - 1] : list;
+        resolve({
+          version: latest.version,
+          dependencies: latest.dependencies,
+          peerDependencies: latest.peerDependencies,
+        });
+      }
+    });
+  });
+}
+
+/**
+ * Returns a profile for specified version.
+ * @param {string} v
+ * @return {Promise<Record<string, string | undefined>>}
+ */
+async function getProfile(v) {
+  switch (v) {
+    case "canary-macos": {
+      const { dependencies, peerDependencies } = await fetchPackageInfo(
+        "react-native-macos@canary"
+      );
+      return {
+        "@react-native-community/cli":
+          dependencies["@react-native-community/cli"],
+        "@react-native-community/cli-platform-android":
+          dependencies["@react-native-community/cli-platform-android"],
+        "@react-native-community/cli-platform-ios":
+          dependencies["@react-native-community/cli-platform-ios"],
+        "hermes-engine": dependencies["hermes-engine"],
+        react: peerDependencies["react"],
+        "react-native": REACT_NATIVE_VERSIONS[v],
+        "react-native-macos": "canary",
+        "react-native-windows": undefined,
+      };
+    }
+
+    case "canary-windows": {
+      const { dependencies, peerDependencies } = await fetchPackageInfo(
+        "react-native-windows@canary"
+      );
+      return {
+        "@react-native-community/cli":
+          dependencies["@react-native-community/cli"],
+        "@react-native-community/cli-platform-android":
+          dependencies["@react-native-community/cli-platform-android"],
+        "@react-native-community/cli-platform-ios":
+          dependencies["@react-native-community/cli-platform-ios"],
+        "hermes-engine": dependencies["hermes-engine"],
+        react: peerDependencies["react"],
+        "react-native": REACT_NATIVE_VERSIONS[v],
+        "react-native-macos": undefined,
+        "react-native-windows": "canary",
+      };
+    }
+
+    case "main": {
+      const { dependencies, peerDependencies } = await fetchPackageInfo(
+        "react-native@nightly"
+      );
+      return {
+        "@react-native-community/cli":
+          dependencies["@react-native-community/cli"],
+        "@react-native-community/cli-platform-android":
+          dependencies["@react-native-community/cli-platform-android"],
+        "@react-native-community/cli-platform-ios":
+          dependencies["@react-native-community/cli-platform-ios"],
+        "hermes-engine": dependencies["hermes-engine"],
+        react: peerDependencies["react"],
+        "react-native": "facebook/react-native",
+        "react-native-macos": undefined,
+        "react-native-windows": undefined,
```

Comment:
small Q: why here it's not
```
        "react-native-macos": "microsoft/react-native-macos",
        "react-native-windows": "microsoft/react-native-windows",
```

Reply:
Historically, react-native-macos and react-native-windows lag behind react-native upstream. Dependencies can quickly become a nightmare if they're perfectly aligned. Here we unset both to be safe. To test react-native-macos or react-native-windows main/nightly, we use canary-macos or canary-windows instead.

Reply:
makes sense!